header/sync/sync.go

```diff
@@ -201,43 +220,48 @@ func (s *Syncer) processIncoming(ctx context.Context, maybeHead *header.Extended
 			"err", err)
 		// might be a storage error or something else, but we can still try to continue processing 'maybeHead'
 	}
-
 	// 2. Get known trusted head, so we can verify maybeHead
 	trstHead, err := s.trustedHead(ctx)
 	if err != nil {
 		log.Errorw("getting trusted head", "err", err)
 		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
 	}
 
-	// 3. Filter out maybeHead if behind trusted
-	if maybeHead.Height <= trstHead.Height {
+	// 3. Attempt to verify maybeHead
+	return s.verifyMaybeHead(trstHead, maybeHead)
+}
+
+// verifyMaybeHead verifies the potential new head against the trusted head, and if
+// the potential new head is valid, sets it as the sync target.
+func (s *Syncer) verifyMaybeHead(trusted, maybe *header.ExtendedHeader) pubsub.ValidationResult {
+	// 1. Filter out maybe if behind trusted
+	if maybe.Height <= trusted.Height {
 		log.Warnw("received known header",
-			"height", maybeHead.Height,
-			"hash", maybeHead.Hash())
+			"height", maybe.Height,
+			"hash", maybe.Hash())
 		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
 	}
 
-	// 4. Verify maybeHead against trusted
-	err = trstHead.VerifyNonAdjacent(maybeHead)
+	// 2. Verify maybeHead against trusted
+	err := trusted.VerifyNonAdjacent(maybe)
 	var verErr *header.VerifyError
 	if errors.As(err, &verErr) {
 		log.Errorw("invalid header",
-			"height_of_invalid", maybeHead.Height,
-			"hash_of_invalid", maybeHead.Hash(),
-			"height_of_trusted", trstHead.Height,
-			"hash_of_trusted", trstHead.Hash(),
+			"height_of_invalid", maybe.Height,
+			"hash_of_invalid", maybe.Hash(),
+			"height_of_trusted", trusted.Height,
+			"hash_of_trusted", trusted.Hash(),
 			"reason", verErr.Reason)
 		return pubsub.ValidationReject
 	}
-
-	// 5. Save verified header to pending cache
+	// 3. Save verified header to pending cache
 	// NOTE: Pending cache can't be DOSed as we verify above each header against a trusted one.
-	s.pending.Add(maybeHead)
-	// and trigger sync to catch-up
+	s.pending.Add(maybe)
+	// 4. And trigger sync to catch up to new sync target
```

Comment:
Hmm, this can cause an endless syncing loop. TrustedHead is triggered by wantSync, and now trustedHead triggers the wantSync throught `verifyMaybeHead`, so there is definitely a cycel.